File: spec/test_tree_ps.c
#include "testparser.h"
#include <assert.h>
#include <string.h>
#include "testutils.h"

int main()
{
    char const * input = "a, ((b)), b";
    p_context_t * context;
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert_eq(P_SUCCESS, p_parse(context));
    PStartS start = p_result(context);
    assert(p_node_valid(p_PStartS_pItems1(start)));
    assert(p_node_valid(p_PStartS_pItems(start)));
    PItemsS items = p_PStartS_pItems(start);
    assert(p_node_valid(p_PItemsS_pItem(items)));
    assert(p_node_valid(p_tree_walk_PItemsS(items, pItem, pToken1)));
    assert_eq(TOKEN_a, p_tree_walk_PItemsS(items, pItem, pToken1, token));
    assert_eq(11, p_tree_walk_PItemsS(items, pItem, pToken1, pvalue));
    assert(p_node_valid(p_PItemsS_pItemsMore(items)));
    PItemsMoreS itemsmore = p_PItemsS_pItemsMore(items);
    assert(p_node_valid(p_PItemsMoreS_pItem(itemsmore)));
    assert(p_node_valid(p_tree_walk_PItemsMoreS(itemsmore, pItem, pItem)));
    assert(p_node_valid(p_tree_walk_PItemsMoreS(itemsmore, pItem, pItem, pItem)));
    assert(p_node_valid(p_tree_walk_PItemsMoreS(itemsmore, pItem, pItem, pItem, pToken1)));
    assert_eq(TOKEN_b, p_tree_walk_PItemsMoreS(itemsmore, pItem, pItem, pItem, pToken1, token));
    assert_eq(22, p_tree_walk_PItemsMoreS(itemsmore, pItem, pItem, pItem, pToken1, pvalue));
    assert(p_node_valid(p_PItemsMoreS_pItemsMore(itemsmore)));
    itemsmore = p_PItemsMoreS_pItemsMore(itemsmore);
    assert(p_node_valid(p_PItemsMoreS_pItem(itemsmore)));
    assert(p_node_valid(p_tree_walk_PItemsMoreS(itemsmore, pItem, pToken1)));
    assert_eq(TOKEN_b, p_tree_walk_PItemsMoreS(itemsmore, pItem, pToken1, token));
    assert_eq(22, p_tree_walk_PItemsMoreS(itemsmore, pItem, pToken1, pvalue));
    assert(!p_node_valid(p_PItemsMoreS_pItemsMore(itemsmore)));

    p_context_delete(context);

    input = "";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert_eq(P_SUCCESS, p_parse(context));
    start = p_result(context);
    assert(!p_node_valid(p_PStartS_pItems(start)));

    p_context_delete(context);

    input = "2 1";
    context = p_context_new((uint8_t const *)input, strlen(input));
    assert_eq(P_SUCCESS, p_parse(context));
    start = p_result(context);
    assert(p_node_valid(p_PStartS_pItems(start)));
    assert(p_node_valid(p_tree_walk_PStartS(start, pItems, pItem)));
    assert(p_node_valid(p_tree_walk_PStartS(start, pItems, pItem, pDual)));
    assert(p_node_valid(p_tree_walk_PStartS(start, pItems, pItem, pDual, pTwo1)));
    assert(p_node_valid(p_tree_walk_PStartS(start, pItems, pItem, pDual, pOne2)));
    assert(!p_node_valid(p_tree_walk_PStartS(start, pItems, pItem, pDual, pTwo2)));
    assert(!p_node_valid(p_tree_walk_PStartS(start, pItems, pItem, pDual, pOne1)));

    p_context_delete(context);

    return 0;
}

File: spec/test_tree_ps.rs
use testparser::*;

fn main() {
    let input = b"a, ((b)), b";
    let mut context = p_context_new(input);
    assert_eq!(P_SUCCESS, p_parse(&mut context));
    {
        let start = p_result(&context);
        assert!(start.pItems1().valid());
        assert!(start.pItems().valid());
        let items = start.pItems();
        assert!(items.pItem().valid());
        assert!(items.pItem().pToken1().valid());
        assert_eq!(TOKEN_a, items.pItem().pToken1().token());
        assert_eq!(11, items.pItem().pToken1().pvalue());
        let itemsmore = items.pItemsMore();
        assert!(itemsmore.pItem().pItem().pItem().pToken1().valid());
        assert_eq!(TOKEN_b, itemsmore.pItem().pItem().pItem().pToken1().token());
        assert_eq!(22, itemsmore.pItem().pItem().pItem().pToken1().pvalue());
        assert!(itemsmore.pItemsMore().valid());
        let itemsmore = itemsmore.pItemsMore();
        assert_eq!(TOKEN_b, itemsmore.pItem().pToken1().token());
        assert!(!itemsmore.pItemsMore().valid());
    }
    p_context_delete(context);

    /* Empty input yields a Start node with no Items child. */
    let mut context = p_context_new(b"");
    assert_eq!(P_SUCCESS, p_parse(&mut context));
    assert!(!p_result(&context).pItems().valid());
    p_context_delete(context);

    /* Dual rule alternative field positions. */
    let mut context = p_context_new(b"2 1");
    assert_eq!(P_SUCCESS, p_parse(&mut context));
    {
        let start = p_result(&context);
        assert!(start.pItems().pItem().pDual().pTwo1().valid());
        assert!(start.pItems().pItem().pDual().pOne2().valid());
        assert!(!start.pItems().pItem().pDual().pTwo2().valid());
        assert!(!start.pItems().pItem().pDual().pOne1().valid());
    }
    p_context_delete(context);

    println!("ok");
}
